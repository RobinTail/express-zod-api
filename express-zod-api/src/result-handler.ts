import { Request, Response } from "express";
import { globalRegistry, z } from "zod/v4";
import {
  ApiResponse,
  defaultStatusCodes,
  NormalizedResponse,
} from "./api-response";
import { FlatObject, isObject } from "./common-helpers";
import { contentTypes } from "./content-type";
import { IOSchema } from "./io-schema";
import { ActualLogger } from "./logger-helpers";
import {
  DiscriminatedResult,
  ensureHttpError,
  getPublicErrorMessage,
  logServerError,
  normalize,
  ResultSchema,
} from "./result-helpers";

type Handler<RES = unknown> = (
  params: DiscriminatedResult & {
    /** null in case of failure to parse or to find the matching endpoint (error: not found) */
    input: FlatObject | null;
    /** can be empty: check presence of the required property using "in" operator */
    options: FlatObject;
    request: Request;
    response: Response<RES>;
    logger: ActualLogger;
  },
) => void | Promise<void>;

export type Result<S extends z.ZodType = z.ZodType> =
  | S // plain schema, default status codes applied
  | ApiResponse<S> // single response definition, status code(s) customizable
  | ApiResponse<S>[]; // Feature #1431: different responses for different status codes (non-empty, prog. check!)

export type LazyResult<R extends Result, A extends unknown[] = []> = (
  ...args: A
) => R;

export abstract class AbstractResultHandler {
  readonly #handler: Handler;
  /** @internal */
  public abstract getPositiveResponse(output: IOSchema): NormalizedResponse[];
  /** @internal */
  public abstract getNegativeResponse(): NormalizedResponse[];
  protected constructor(handler: Handler) {
    this.#handler = handler;
  }
  public execute(...params: Parameters<Handler>) {
    return this.#handler(...params);
  }
}

export class ResultHandler<
  POS extends Result,
  NEG extends Result,
> extends AbstractResultHandler {
  readonly #positive: POS | LazyResult<POS, [IOSchema]>;
  readonly #negative: NEG | LazyResult<NEG>;

  constructor(params: {
    /** @desc A description of the API response in case of success (schema, status code, MIME type) */
    positive: POS | LazyResult<POS, [IOSchema]>;
    /** @desc A description of the API response in case of error (schema, status code, MIME type) */
    negative: NEG | LazyResult<NEG>;
    /** @desc The actual implementation to transmit the response in any case */
    handler: Handler<z.output<ResultSchema<POS> | ResultSchema<NEG>>>;
  }) {
    super(params.handler);
    this.#positive = params.positive;
    this.#negative = params.negative;
  }

  /** @internal */
  public override getPositiveResponse(output: IOSchema) {
    return normalize(this.#positive, {
      variant: "positive",
      args: [output],
      statusCodes: [defaultStatusCodes.positive],
      mimeTypes: [contentTypes.json],
    });
  }

  /** @internal */
  public override getNegativeResponse() {
    return normalize(this.#negative, {
      variant: "negative",
      args: [],
      statusCodes: [defaultStatusCodes.negative],
      mimeTypes: [contentTypes.json],
    });
  }
}

export const defaultResultHandler = new ResultHandler({
  positive: (output) => {
    const responseSchema = z.object({
      status: z.literal("success"),
      data: output,
    });
    const { examples = [] } = globalRegistry.get(output) || {}; // pulling down:
    if (examples.length) {
      globalRegistry.add(responseSchema, {
        examples: examples.map((data) => ({
          status: "success" as const,
          data,
        })),
      });
    }
    return responseSchema;
  },
  negative: z
    .object({
      status: z.literal("error"),
      error: z.object({ message: z.string() }),
    })
    .example({
      status: "error",
      error: { message: "Sample error message" },
    }),
  handler: ({ error, input, output, request, response, logger }) => {
    if (error) {
      const httpError = ensureHttpError(error);
      logServerError(httpError, logger, request, input);
      return void response
        .status(httpError.statusCode)
        .set(httpError.headers)
        .json({
          status: "error",
          error: { message: getPublicErrorMessage(httpError) },
        });
    }
    response
      .status(defaultStatusCodes.positive)
      .json({ status: "success", data: output });
  },
});

/**
 * @deprecated Resist the urge of using it: this handler is designed only to simplify the migration of legacy APIs.
 * @desc Responding with array is a bad practice keeping your endpoints from evolving without breaking changes.
 * @desc This handler expects your endpoint to have the property 'items' in the output object schema
 * */
export const arrayResultHandler = new ResultHandler({
  positive: (output) => {
    const responseSchema =
      output instanceof z.ZodObject &&
      "items" in output.shape &&
      output.shape.items instanceof z.ZodArray
        ? output.shape.items
        : z.array(z.any());
    const meta = responseSchema.meta();
    if (meta?.examples?.length) return responseSchema; // has examples on the items, or pull down:
    const examples = (globalRegistry.get(output)?.examples || [])
      .filter(
        (example): example is { items: unknown[] } =>
          isObject(example) &&
          "items" in example &&
          Array.isArray(example.items),
      )
      .map((example) => example.items);
    if (examples.length) {
      globalRegistry
        .remove(responseSchema) // reassign to avoid cloning
        .add(responseSchema, { ...meta, examples });
    }
    return responseSchema;
  },
  negative: z.string().example("Sample error message"),
  handler: ({ response, output, error, logger, request, input }) => {
    if (error) {
      const httpError = ensureHttpError(error);
      logServerError(httpError, logger, request, input);
      return void response
        .status(httpError.statusCode)
        .type("text/plain")
        .send(getPublicErrorMessage(httpError));
    }
    if ("items" in output && Array.isArray(output.items)) {
      return void response
        .status(defaultStatusCodes.positive)
        .json(output.items);
    }
    throw new Error("Property 'items' is missing in the endpoint output");
  },
});
