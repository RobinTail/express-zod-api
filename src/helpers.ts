import {Request} from 'express';
import {z} from 'zod';
import {MiddlewareDefinition} from './middleware';

export type FlatObject = Record<string, any>;
export type ObjectSchema = z.AnyZodObject;

export type Merge<A extends ObjectSchema, B extends ObjectSchema | any> = z.ZodObject<
  // eslint-disable-next-line @typescript-eslint/ban-types
  A['shape'] & (B extends ObjectSchema ? B['shape'] : {}),
  A['_unknownKeys'],
  A['_catchall']
>;

export function combineEndpointAndMiddlewareInputSchemas<IN extends ObjectSchema, mIN>(
  input: IN,
  middlewares: MiddlewareDefinition<any, any, any>[]
): Merge<IN, mIN> {
  if (middlewares.length === 0) {
    return input as any as Merge<IN, mIN>;
  }
  return middlewares
    .map((middleware) => middleware.input)
    .reduce((carry: ObjectSchema, schema) => carry.merge(schema))
    .merge(input) as Merge<IN, mIN>;
}

export function getInitialInput(request: Request): any {
  switch (request.method) {
    case 'POST':
    case 'PUT':
    case 'PATCH':
      return request.body;
    case 'GET':
      return request.query;
    case 'DELETE': // _may_ have body
    default:
      return {...request.query, ...request.body};
  }
}
