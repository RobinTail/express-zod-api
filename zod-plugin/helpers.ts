import type { z } from "zod";
import { unpack } from "./packer";

export const getBrand = (subject: z.core.$ZodType) => {
  const { brand } = unpack(subject) || {};
  if (
    typeof brand === "symbol" ||
    typeof brand === "string" ||
    typeof brand === "number"
  )
    return brand;
  return undefined;
};
