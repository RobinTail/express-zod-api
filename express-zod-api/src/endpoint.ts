import { Request, Response } from "express";
import { z } from "zod";
import { NormalizedResponse, ResponseVariant } from "./api-response";
import { hasRaw, hasUpload } from "./deep-checks";
import {
  FlatObject,
  getActualMethod,
  getInput,
  ensureError,
} from "./common-helpers";
import { CommonConfig } from "./config-type";
import {
  InputValidationError,
  OutputValidationError,
  ResultHandlerError,
} from "./errors";
import { IOSchema } from "./io-schema";
import { lastResortHandler } from "./last-resort";
import { ActualLogger } from "./logger-helpers";
import { LogicalContainer } from "./logical-container";
import { AuxMethod, Method } from "./method";
import { AbstractMiddleware, ExpressMiddleware } from "./middleware";
import { ContentType } from "./content-type";
import { Nestable } from "./routable";
import { AbstractResultHandler } from "./result-handler";
import { Security } from "./security";

export type Handler<IN, OUT, OPT> = (params: {
  input: IN;
  options: OPT;
  logger: ActualLogger;
}) => Promise<OUT>;

type DescriptionVariant = "short" | "long";
type IOVariant = "input" | "output";

export abstract class AbstractEndpoint extends Nestable {
  public abstract execute(params: {
    request: Request;
    response: Response;
    logger: ActualLogger;
    config: CommonConfig;
  }): Promise<void>;
  public abstract getDescription(
    variant: DescriptionVariant,
  ): string | undefined;
  public abstract getMethods(): ReadonlyArray<Method> | undefined;
  public abstract getSchema(variant: IOVariant): IOSchema;
  public abstract getResponses(
    variant: ResponseVariant,
  ): ReadonlyArray<NormalizedResponse>;
  // @todo should return ReadonlyArray
  public abstract getSecurity(): LogicalContainer<Security>[];
  public abstract getScopes(): ReadonlyArray<string>;
  public abstract getTags(): ReadonlyArray<string>;
  public abstract getOperationId(method: Method): string | undefined;
  public abstract getRequestType(): ContentType;
}

export class Endpoint<
  IN extends IOSchema,
  OUT extends IOSchema,
  OPT extends FlatObject,
> extends AbstractEndpoint {
  readonly #descriptions: Record<DescriptionVariant, string | undefined>;
  readonly #methods?: ReadonlyArray<Method>;
  readonly #middlewares: AbstractMiddleware[];
  readonly #responses: Record<
    ResponseVariant,
    ReadonlyArray<NormalizedResponse>
  >;
  readonly #handler: Handler<z.output<IN>, z.input<OUT>, OPT>;
  readonly #resultHandler: AbstractResultHandler;
  readonly #schemas: { input: IN; output: OUT };
  readonly #scopes: ReadonlyArray<string>;
  readonly #tags: ReadonlyArray<string>;
  readonly #getOperationId: (method: Method) => string | undefined;
  readonly #requestType: ContentType;

  constructor({
    methods,
    inputSchema,
    outputSchema,
    handler,
    resultHandler,
    getOperationId = () => undefined,
    scopes = [],
    middlewares = [],
    tags = [],
    description: long,
    shortDescription: short,
  }: {
    middlewares?: AbstractMiddleware[];
    inputSchema: IN;
    outputSchema: OUT;
    handler: Handler<z.output<IN>, z.input<OUT>, OPT>;
    resultHandler: AbstractResultHandler;
    description?: string;
    shortDescription?: string;
    getOperationId?: (method: Method) => string | undefined;
    methods?: Method[];
    scopes?: string[];
    tags?: string[];
  }) {
    super();
    this.#handler = handler;
    this.#resultHandler = resultHandler;
    this.#middlewares = middlewares;
    this.#getOperationId = getOperationId;
    this.#methods = Object.freeze(methods);
    this.#scopes = Object.freeze(scopes);
    this.#tags = Object.freeze(tags);
    this.#descriptions = { long, short };
    this.#schemas = { input: inputSchema, output: outputSchema };
    this.#responses = {
      positive: Object.freeze(resultHandler.getPositiveResponse(outputSchema)),
      negative: Object.freeze(resultHandler.getNegativeResponse()),
    };
    this.#requestType = hasUpload(inputSchema)
      ? "upload"
      : hasRaw(inputSchema)
        ? "raw"
        : "json";
  }

  public override clone() {
    return new Endpoint({
      // @todo internal props may be unfrozen, but methods returning them can freeze
      methods: this.#methods?.slice(),
      inputSchema: this.#schemas.input,
      outputSchema: this.#schemas.output,
      handler: this.#handler,
      resultHandler: this.#resultHandler,
      getOperationId: this.#getOperationId,
      scopes: this.#scopes.slice(),
      middlewares: this.#middlewares,
      tags: this.#tags.slice(),
      description: this.#descriptions.long,
      shortDescription: this.#descriptions.short,
    }) as this;
  }

  public override getDescription(variant: DescriptionVariant) {
    return this.#descriptions[variant];
  }

  public override getMethods() {
    return this.#methods;
  }

  public override getSchema(variant: "input"): IN;
  public override getSchema(variant: "output"): OUT;
  public override getSchema(variant: IOVariant) {
    return this.#schemas[variant];
  }

  public override getRequestType() {
    return this.#requestType;
  }

  public override getResponses(variant: ResponseVariant) {
    return this.#responses[variant];
  }

  public override getSecurity() {
    return this.#middlewares
      .map((middleware) => middleware.getSecurity())
      .filter((entry) => entry !== undefined);
  }

  public override getScopes() {
    return this.#scopes;
  }

  public override getTags() {
    return this.#tags;
  }

  public override getOperationId(method: Method): string | undefined {
    return this.#getOperationId(method);
  }

  async #parseOutput(output: z.input<OUT>) {
    try {
      return (await this.#schemas.output.parseAsync(output)) as FlatObject;
    } catch (e) {
      throw e instanceof z.ZodError ? new OutputValidationError(e) : e;
    }
  }

  async #runMiddlewares({
    method,
    logger,
    options,
    response,
    ...rest
  }: {
    method: Method | AuxMethod;
    input: Readonly<FlatObject>; // Issue #673: input is immutable, since this.inputSchema is combined with ones of middlewares
    request: Request;
    response: Response;
    logger: ActualLogger;
    options: Partial<OPT>;
  }) {
    for (const mw of this.#middlewares) {
      if (method === "options" && !(mw instanceof ExpressMiddleware)) continue;
      Object.assign(
        options,
        await mw.execute({ ...rest, options, response, logger }),
      );
      if (response.writableEnded) {
        logger.warn(
          "A middleware has closed the stream. Accumulated options:",
          options,
        );
        break;
      }
    }
  }

  async #parseAndRunHandler({
    input,
    ...rest
  }: {
    input: Readonly<FlatObject>;
    options: OPT;
    logger: ActualLogger;
  }) {
    let finalInput: z.output<IN>; // final input types transformations for handler
    try {
      finalInput = (await this.#schemas.input.parseAsync(
        input,
      )) as z.output<IN>;
    } catch (e) {
      throw e instanceof z.ZodError ? new InputValidationError(e) : e;
    }
    return this.#handler({ ...rest, input: finalInput });
  }

  async #handleResult({
    error,
    ...rest
  }: {
    error: Error | null;
    request: Request;
    response: Response;
    logger: ActualLogger;
    input: FlatObject;
    output: FlatObject | null;
    options: Partial<OPT>;
  }) {
    try {
      await this.#resultHandler.execute({ ...rest, error });
    } catch (e) {
      lastResortHandler({
        ...rest,
        error: new ResultHandlerError(ensureError(e), error || undefined),
      });
    }
  }

  public override async execute({
    request,
    response,
    logger,
    config,
  }: {
    request: Request;
    response: Response;
    logger: ActualLogger;
    config: CommonConfig;
  }) {
    const method = getActualMethod(request);
    const options: Partial<OPT> = {};
    let output: FlatObject | null = null;
    let error: Error | null = null;
    const input = getInput(request, config.inputSources);
    try {
      await this.#runMiddlewares({
        method,
        input,
        request,
        response,
        logger,
        options,
      });
      if (response.writableEnded) return;
      if (method === "options") return void response.status(200).end();
      output = await this.#parseOutput(
        await this.#parseAndRunHandler({
          input,
          logger,
          options: options as OPT, // ensured the complete OPT by writableEnded condition and try-catch
        }),
      );
    } catch (e) {
      error = ensureError(e);
    }
    await this.#handleResult({
      input,
      output,
      request,
      response,
      error,
      logger,
      options,
    });
  }
}
