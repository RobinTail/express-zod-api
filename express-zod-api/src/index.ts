import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
