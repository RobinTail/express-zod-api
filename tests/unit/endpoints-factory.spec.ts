import { createMiddleware, EndpointsFactory, z } from "../../src";
import { Endpoint } from "../../src/endpoint";
import { ResultHandlerDefinition } from "../../src/result-handler";
import { expectType } from "tsd";
import { serializeSchemaForTest } from "../helpers";

describe("EndpointsFactory", () => {
  /* eslint-disable @typescript-eslint/dot-notation */

  describe(".constructor()", () => {
    test("Should create the empty factory with result handler", () => {
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      );
      expect(factory).toBeInstanceOf(EndpointsFactory);
      expect(factory["middlewares"]).toStrictEqual([]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });

    test("Should create the factory with middleware and result handler", () => {
      const middleware = createMiddleware({
        input: z.object({
          n: z.number(),
        }),
        middleware: jest.fn(),
      });
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      ).addMiddleware(middleware);
      expect(factory["middlewares"]).toStrictEqual([middleware]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });
  });

  describe(".addMiddleware()", () => {
    test("Should create a new factory with a middleware and the same result handler", () => {
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      );
      const middleware = createMiddleware({
        input: z.object({
          n: z.number(),
        }),
        middleware: jest.fn(),
      });
      const newFactory = factory.addMiddleware(middleware);
      expect(factory["middlewares"]).toStrictEqual([]);
      expect(factory["resultHandler"]).toStrictEqual(resultHandlerMock);
      expect(newFactory["middlewares"]).toStrictEqual([middleware]);
      expect(newFactory["resultHandler"]).toStrictEqual(resultHandlerMock);
    });
  });

  describe(".build()", () => {
    test("Should create an endpoint with simple middleware", () => {
      const middleware = createMiddleware({
        input: z.object({
          n: z.number(),
        }),
        middleware: jest.fn(),
      });
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      ).addMiddleware(middleware);
      const handlerMock = jest.fn();
      const endpoint = factory.build({
        method: "get",
        input: z.object({
          s: z.string(),
        }),
        output: z.object({
          b: z.boolean(),
        }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.getMethods()).toStrictEqual(["get"]);
      expect(endpoint["middlewares"]).toStrictEqual([middleware]);
      expect(serializeSchemaForTest(endpoint["inputSchema"])).toMatchSnapshot();
      expect(
        serializeSchemaForTest(endpoint["outputSchema"])
      ).toMatchSnapshot();
      expect(endpoint["handler"]).toStrictEqual(handlerMock);
      expect(endpoint["resultHandler"]).toStrictEqual(resultHandlerMock);
      expectType<{
        n: z.ZodNumber;
        s: z.ZodString;
      }>(endpoint["inputSchema"].shape);
    });

    test("Should create an endpoint with intersection middleware", () => {
      const middleware = createMiddleware({
        input: z
          .object({
            n1: z.number(),
          })
          .and(
            z.object({
              n2: z.number(),
            })
          ),
        middleware: jest.fn(),
      });
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      ).addMiddleware(middleware);
      const handlerMock = jest.fn();
      const endpoint = factory.build({
        methods: ["get"],
        input: z.object({
          s: z.string(),
        }),
        output: z.object({
          b: z.boolean(),
        }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.getMethods()).toStrictEqual(["get"]);
      expect(endpoint["middlewares"]).toStrictEqual([middleware]);
      expect(serializeSchemaForTest(endpoint["inputSchema"])).toMatchSnapshot();
      expect(
        serializeSchemaForTest(endpoint["outputSchema"])
      ).toMatchSnapshot();
      expect(endpoint["handler"]).toStrictEqual(handlerMock);
      expect(endpoint["resultHandler"]).toStrictEqual(resultHandlerMock);
      expectType<{
        n1: z.ZodNumber;
        n2: z.ZodNumber;
        s: z.ZodString;
      }>(endpoint["inputSchema"].shape);
    });

    test("Should create an endpoint with union middleware", () => {
      const middleware = createMiddleware({
        input: z
          .object({
            n1: z.number(),
          })
          .or(
            z.object({
              n2: z.number(),
            })
          ),
        middleware: jest.fn(),
      });
      const resultHandlerMock = { handler: jest.fn() };
      const factory = new EndpointsFactory(
        resultHandlerMock as unknown as ResultHandlerDefinition<any, any>
      ).addMiddleware(middleware);
      const handlerMock = jest.fn().mockImplementation((params) => ({
        input: params.input,
        b: true,
      }));
      const endpoint = factory.build({
        methods: ["get"],
        input: z.object({
          s: z.string(),
        }),
        output: z.object({
          b: z.boolean(),
        }),
        handler: handlerMock,
      });
      expect(endpoint).toBeInstanceOf(Endpoint);
      expect(endpoint.getMethods()).toStrictEqual(["get"]);
      expect(endpoint["middlewares"]).toStrictEqual([middleware]);
      expect(serializeSchemaForTest(endpoint["inputSchema"])).toMatchSnapshot();
      expect(
        serializeSchemaForTest(endpoint["outputSchema"])
      ).toMatchSnapshot();
      expect(endpoint["handler"]).toStrictEqual(handlerMock);
      expect(endpoint["resultHandler"]).toStrictEqual(resultHandlerMock);
      expectType<
        (
          | {
              n1: z.ZodNumber;
            }
          | {
              n2: z.ZodNumber;
            }
        ) & {
          s: z.ZodString;
        }
      >(endpoint["inputSchema"].shape);
      expectType<{
        n1?: z.ZodNumber;
        n2?: z.ZodNumber;
        s: z.ZodString;
      }>(endpoint["inputSchema"].shape);
    });
  });
});
