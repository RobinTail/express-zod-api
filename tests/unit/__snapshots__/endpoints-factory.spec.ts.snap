// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EndpointsFactory .build() Should create an endpoint with intersection middleware 1`] = `
Object {
  "n1": ZodNumber {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "int": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "negative": [Function],
    "nonnegative": [Function],
    "nonpositive": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "positive": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "n2": ZodNumber {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "int": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "negative": [Function],
    "nonnegative": [Function],
    "nonpositive": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "positive": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "s": ZodString {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "_regex": [Function],
    "array": [Function],
    "default": [Function],
    "email": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "nonempty": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "regex": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "url": [Function],
    "uuid": [Function],
  },
}
`;

exports[`EndpointsFactory .build() Should create an endpoint with intersection middleware 2`] = `
Object {
  "b": ZodBoolean {
    "_def": Object {},
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
}
`;

exports[`EndpointsFactory .build() Should create an endpoint with simple middleware 1`] = `
Object {
  "n": ZodNumber {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "int": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "negative": [Function],
    "nonnegative": [Function],
    "nonpositive": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "positive": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "s": ZodString {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "_regex": [Function],
    "array": [Function],
    "default": [Function],
    "email": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "nonempty": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "regex": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "url": [Function],
    "uuid": [Function],
  },
}
`;

exports[`EndpointsFactory .build() Should create an endpoint with simple middleware 2`] = `
Object {
  "b": ZodBoolean {
    "_def": Object {},
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
}
`;

exports[`EndpointsFactory .build() Should create an endpoint with union middleware 1`] = `
Object {
  "n1": ZodOptional {
    "_def": Object {
      "innerType": ZodNumber {
        "_def": Object {
          "checks": Array [],
        },
        "_parseInternalOptionalParams": [Function],
        "_parseWithInvalidFallback": [Function],
        "array": [Function],
        "default": [Function],
        "int": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "max": [Function],
        "min": [Function],
        "negative": [Function],
        "nonnegative": [Function],
        "nonpositive": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "positive": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "n2": ZodOptional {
    "_def": Object {
      "innerType": ZodNumber {
        "_def": Object {
          "checks": Array [],
        },
        "_parseInternalOptionalParams": [Function],
        "_parseWithInvalidFallback": [Function],
        "array": [Function],
        "default": [Function],
        "int": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "max": [Function],
        "min": [Function],
        "negative": [Function],
        "nonnegative": [Function],
        "nonpositive": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "positive": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
      },
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
  "s": ZodString {
    "_def": Object {
      "checks": Array [],
    },
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "_regex": [Function],
    "array": [Function],
    "default": [Function],
    "email": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "max": [Function],
    "min": [Function],
    "nonempty": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "regex": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "url": [Function],
    "uuid": [Function],
  },
}
`;

exports[`EndpointsFactory .build() Should create an endpoint with union middleware 2`] = `
Object {
  "b": ZodBoolean {
    "_def": Object {},
    "_parseInternalOptionalParams": [Function],
    "_parseWithInvalidFallback": [Function],
    "array": [Function],
    "default": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
  },
}
`;
